proto/cachepubsub.proto

```diff
@@ -0,0 +1,126 @@
+option csharp_namespace = "Momento.Protos.CacheClient.Pubsub";
+
+package cache_client.pubsub;
+
+// For working with topics in a cache.
+// Momento topics are conceptually located on a cache. They are best-effort multicast.
+// To use them, create a cache then start subscribing and publishing!
+//
+// Momento topic subscriptions try to give you information about the quality of the
+//   stream you are receiving. For example, you might miss messages if your network
+//   is slow, or if some intermediate switch fails, or due to rate limiting. It is
+//   also possible, though we try to avoid it, that messages could briefly come out
+//   of order between subscribers.
+//   We try to tell you when things like this happen via a Discontinuity in your
+//   subscription stream. If you do not care about occasional discontinuities then
+//   don't bother handling them! You might still want to log them just in case ;-)
+service Pubsub {
+  // Publish a message to a topic.
+  //
+  // If a topic has no subscribers, then the effect of Publish MAY be either of:
+  // * It is dropped and the topic is nonexistent.
+  // * It is accepted to the topic as the next message.
+  //
+  // Publish() does not wait for subscribers to accept. It returns Ok upon accepting
+  // the topic value. It also returns Ok if there are no subscribers and the value
+  // happens to be dropped. Publish() can not guarantee delivery in theory but in
+  // practice it should almost always deliver to subscribers.
+  //
+  // REQUIRES HEADER authorization: Momento auth token
+  rpc Publish(_PublishRequest) returns (_Empty);
+
+  // Subscribe to notifications from a topic.
+  //
+  // You will receive a stream of values and (hopefully occasional) discontinuities.
+  // Values will appear as copies of the payloads you Publish() to the topic.
+  //
+  // REQUIRES HEADER authorization: Momento auth token
+  rpc Subscribe(_SubscriptionRequest) returns (stream _SubscriptionItem);
+}
+
+message _Empty {}
+
+// A value to publish through a topic.
+message _PublishRequest {
+  // Cache namespace for the topic to which you want to send the value.
+  string cache_name = 1;
+  // The literal topic name to which you want to send the value.
+  string topic = 2;
+  // The value you want to send to the topic. All current subscribers will receive
+  // this, should the whims of the Internet prove merciful.
+  _TopicValue value = 3;
+}
+
+// A description of how you want to subscribe to a topic.
+message _SubscriptionRequest {
+  // Cache namespace for the topic to which you want to subscribe.
+  string cache_name = 1;
+
+  // The literal topic name to which you want to subscribe.
+  string topic = 2;
+
+  // --> Providing this is not required. <--
+  //
+  // If provided, attempt to reconnect to the topic and replay messages starting from
+  // the provided sequence number. You may get a discontinuity if some (or all) of the
+  // messages are not available.
```

Comment:
based on what you told me about how you plan to manage the messages in memory, this seems like it would rarely ever return many missed messages.  What am I missing?

(In theory I like this as part of the API design but it just doesn't seem like it would provide much value in most cases?)

Reply:
Our initial implementation will not have an in-memory topic buffer. Any deployment to the front-end service will cause a reconnect, for which you will almost surely receive a discontinuity. You do not need to do anything about it today, but I'd recommend at least logging so we can more easily troubleshoot issues (e.g., is your subscription simply not working or are you getting a ton of gaps in your stream?)

Similarly, a backend deployment will cause a relocation of the Topic's source of truth. The coordination of a happy-path quiesce point and resume is not going to fit in our early milestones either.

Discontinuities are a good way for users to consider whether momento topics are giving them a good enough quality of subscription relative to other potential lossy pubsub services (or applications), and a good way for us to measure when we're doing well and when we could improve.

Reply:
To be fair, deployments of either service role should be pretty rare relative to a popular Topic's event rate; and when nothing is broken or relocating there should be no discontinuities.

Reply:
i haven't spent enough time thinking about deployments.  we should talk at some point about what the implications of this are in terms of our ability to deploy mr2 and ms2 and if there are any design decisions we can make that will help us maintain a relatively easy story for CD.